src/views/Game/AntiGrief.tsx

```diff
@@ -0,0 +1,283 @@
+    const [clock, setClock] = React.useState<JGOFClockWithTransmitting>(goban?.last_clock as any);
+    const [expiration, setExpiration] = React.useState<number>(null);
+    const [show, setShow] = React.useState(false);
+
+    React.useEffect(() => {
+        setShow(false);
+        setClock(goban?.last_clock as any);
+        goban.on("clock", setClock);
+
+        return () => {
+            goban.off("clock", setClock);
+        };
+    }, [goban, setClock]);
+
+    React.useEffect(() => {
+        setPhase(goban?.engine?.phase);
+    }, [goban?.engine?.phase]);
+
+    React.useEffect(() => {
+        const handleAutoResign = (data?: { player_id: number; expiration: number }) => {
+            setShow(false);
+            setExpiration(data?.expiration);
+        };
+        const handleClearAutoResign = () => {
+            setShow(false);
+            setExpiration(null);
+        };
+
+        goban.on("auto-resign", handleAutoResign);
+        goban.on("clear-auto-resign", handleClearAutoResign);
+
+        return () => {
+            goban.off("auto-resign", handleAutoResign);
+            goban.off("clear-auto-resign", handleClearAutoResign);
+        };
+    }, [goban]);
+
+    React.useEffect(() => {
+        if (expiration) {
+            const timer = setTimeout(() => {
+                setExpiration(null);
+                setShow(true);
+            }, 30 * 1000);
+            return () => {
+                clearTimeout(timer);
+            };
+        }
+    }, [expiration]);
+
+    if (phase !== "play") {
+        return null;
+    }
+
+    const time_control: JGOFTimeControl = goban?.engine?.time_control;
+
+    if (!time_control) {
+        return null;
+    }
+
+    if (time_control.speed === "correspondence") {
+        return null;
+    }
+
+    if (!clock || clock.pause_state) {
+        return null;
+    }
+
+    if (
+        user.id !== goban?.engine?.config?.black_player_id &&
+        user.id !== goban?.engine?.config?.white_player_id &&
+        !user.is_moderator
+    ) {
+        return null;
+    }
+
+    if (!show) {
+        return null;
+    }
+
+    const my_color = user.id === goban?.engine?.config?.black_player_id ? "black" : "white";
+
+    return (
+        <Card className="AntiEscaping">
+            <div>
+                {pgettext(
+                    "This message is shown when one player has left the game",
+                    "Your opponent is no longer connected. You can wait and see if they come back, or end the game.",
+                )}
+            </div>
+            <div>
+                <button
+                    className="danger"
+                    onClick={() => goban?.sendPreventEscaping(my_color, false)}
+                >
+                    {pgettext(
+                        "This button is shown when one player has left the game, it allows the other player to end the game and claim victory",
+                        "End game",
```

Comment:
```suggestion
                        "Claim victory",
```

I think lichess uses the button title Claim Victory, which I think makes it more clear how the result will be counted if you press it, but "End game" is fine too if it's clarified in the message above by saying "or end the game and claim victory."

Reply:
Yeah that's better, thanks!